contracts/tokens/extension/ExtendableDiamond.sol

```diff
@@ -0,0 +1,260 @@
+    */
+    function _registerExtension(address extension) internal virtual returns (bool) {
+        MappedExtensions storage extLibStorage = _extensionStorage();
+        require(extLibStorage.extensions[extension].state == ExtensionState.EXTENSION_NOT_EXISTS, "The extension must not already exist");
+
+        //Interfaces has been validated, lets begin setup
+
+        IExtension ext = IExtension(extension);
+
+        //Next lets figure out what external functions to register in the Extension
+        bytes4[] memory externalFunctions = ext.externalFunctions();
+
+        IDiamondCut.FacetCut[] memory _diamondCut = new IDiamondCut.FacetCut[](1);
+        _diamondCut[0] = IDiamondCut.FacetCut(
+            extension,
+            IDiamondCut.FacetCutAction.Add,
+            externalFunctions
+        );
+        bytes memory initCalldata = abi.encodePacked(abi.encodeWithSelector(IExtension.initialize.selector), extension);
+        LibDiamond.diamondCut(_diamondCut, extension, initCalldata);
+
+        //Finally, add it to storage
+        extLibStorage.extensions[extension] = ExtensionData(
+            ExtensionState.EXTENSION_ENABLED,
+            extLibStorage.registeredExtensions.length,
+            address(this),
+            externalFunctions
+        );
+
+        extLibStorage.registeredExtensions.push(extension);
+        return true;
+    }
+
+    function _upgradeExtension(address extension, address newExtension) internal returns (bool) {
+        MappedExtensions storage extLibStorage = _extensionStorage();
+        require(extLibStorage.extensions[extension].state != ExtensionState.EXTENSION_NOT_EXISTS, "The extension must already exist");
+
+        IExtension ext = IExtension(extension);
+        IExtension newExt = IExtension(newExtension);
+
+        address currentDeployer = ext.extensionDeployer();
+        address newDeployer = newExt.extensionDeployer();
+
+        require(currentDeployer == newDeployer, "Deployer address for new extension is different than current");
+
+        bytes32 currentPackageHash = ext.packageHash();
+        bytes32 newPackageHash = newExt.packageHash();
+
+        require(currentPackageHash == newPackageHash, "Package for new extension is different than current");
+
+        uint256 currentVersion = ext.version();
+        uint256 newVersion = newExt.version();
+
+        require(currentVersion != newVersion, "Versions should not match");
+
+        TokenStandard standard = IToken(address(this)).tokenStandard();
+
+        require(ext.isTokenStandardSupported(standard), "Token standard is not supported in new extension");
+
+        bytes32 interfaceLabel = keccak256(abi.encodePacked(ext.interfaceLabel()));
+        bytes32 newInterfaceLabel = keccak256(abi.encodePacked(newExt.interfaceLabel()));
+
+        require(interfaceLabel == newInterfaceLabel, "Interface labels do not match");
+        
+        _removeExtension(extension);
+
+        _registerExtension(newExtension);
```

Comment:
For this upgrade scenario, would it be worth refactoring to have a single call to diamond cut with a replace action, instead of the current two calls which perform a remove and add separately? Not sure if it would achieve or warrant  any gas savings.